IonDotnet/Internals/Tree/SystemTreeReader.cs

```diff
@@ -0,0 +1,323 @@
+                int newlen = oldlen * 2;
+                Object[] temp = new Object[newlen];
+                Array.Copy(_stack, 0, temp, 0, oldlen);
+                _stack = temp;
+            }
+            _stack[_top++] = _parent;
+            _stack[_top++] = _iter;
+        }
+
+        private void Pop()
+        {
+            _top--;
+            _iter = (IEnumerator<IIonValue>)_stack[_top];
+            _stack[_top] = null;  // Allow iterator to be garbage collected!
+
+            _top--;
+            _parent = (IIonValue)_stack[_top];
+            _stack[_top] = null;
+
+            // We don't know if we're at the end of the container, so check again.
+            _eof = false;
+        }
+
+        public virtual bool HasNext()
+        {
+            IonType next_type = NextHelperSystem();
+            return (next_type != IonType.Null);
+        }
+
+        protected IonType NextHelperSystem()
+        {
+            if (_eof) return IonType.Null;
+            if (_next != null) return _next.Type();
+
+            while(_iter != null && _iter.MoveNext())
+            {
+                _next = _iter.Current;
+            }
+
+            if ((_eof = (_next == null)))
+            {
+                return IonType.Null;
+            }
+            return _next.Type();
+        }
+
+        public void StepOut()
+        {
+            if (_top < 1)
+            {
+                throw new IonException("Cannot stepOut any further, already at top level.");
+            }
+            Pop();
+            _current = null;
+        }
+
+        public string StringValue()
+        {
+            return _current.StringValue;
+        }
+
+        public SymbolToken SymbolValue()
+        {
+            return _current.SymbolValue;
+        }
+
+        public Timestamp TimestampValue()
+        {
+            return _current.TimestampValue;
+        }
+
+        internal class Children : IEnumerator<IIonValue>
+        {
+            bool _eof;
+            int _next_idx;
+            IIonValue _parent;
+            IIonValue _curr;
+
+            public Children(IIonValue parent)
+            {
+                _parent = parent;
+                _next_idx = 0;
+                _curr = null;
+                if (_parent.IsNull)
+                {
+                    // otherwise the empty contents member will cause trouble
+                    _eof = true;
+                }
+            }
+
+            public IIonValue Current
+            {
+                get
+                {
+                    try
+                    {
+                        return _parent.GetElementAt(_next_idx);
+                    }
+                    catch (IndexOutOfRangeException)
+                    {
+                        throw new InvalidOperationException();
+                    }
+                }
+            }
+
+            object IEnumerator.Current => _curr;
```

Comment:
What does this do?

Reply:
Implementations of  ```interface IEnumerator<out T>``` should support simple iteration over a generic collection.